### ISSUE TITLE
Java codegen support for interfaces

### ISSUE THREAD

**Issue by cocreature** [digital-asset/daml#11350]
Depends on #13036

* [ ] exercising inherited interface choices just as if they were directly defined on implementing templates
* [ ] exercising interface choices by interface contract ID
* for every interface type `I`,
  * [ ] a Java type for `ContractId I`, used where that type occurs in any serializable type
  * for every `I`-implementing template `T`,
    * [ ] a convenient function `ContractId T -> ContractId I`
    * [ ] a function `ContractId I -> ContractId T`, with the clear caveat that this conversion is not checked against the ledger
    * the above functions generated as part of `T`, since no `I` knows all its possible `T`s

**Comment by S11001001** [comment]
The presumption of #13036 will be that Java codegen will define a _nominal_, uninhabited-in-Java type for each interface, containing an _inhabited_ `ContractId` type so that interface contract IDs have proper types where they appear in contract data and choices.

As interface contract IDs are serializable, but the interface type itself is not (I presume).

**Comment by S11001001** [comment]
We considered having `T.ContractId <: I.ContractId` for all templates `T` that implement interfaces `I`. However, there's a big problem with this:

Suppose `icid: I.ContractId`. Then `icid instanceof T.ContractId => true` _implies_ that fetching `icid` will yield a `T` contract assuming normal use of a single ledger. However, `icid instanceof T.ContractId => false` **does not** imply that fetching `icid` _will not_ yield a `T` contract. That is because, while `icid` _may_ have arisen from widening, it may alternatively have come from decoding a choice exercise or contract payload, which will not have the embedded information that the Java class `T.ContractId` should be constructed.

That said, it does seem important to have the conversion functions both ways, just not necessarily embedded in the type system.

**Comment by S11001001** [comment]
@realvictorprm Does #13366 support

> a function `ContractId I -> ContractId T`, with the clear caveat that this conversion is not checked against the ledger

?

**Comment by realvictorprm** [comment]
That one is missing, I completely forgot about the checklist 🤦‍♂️however
do we really want this conversion to be available?

Stephen Compall ***@***.***> schrieb am Di., 29. März 2022,
15:31:

**Comment by S11001001** [comment]
> do we really want this conversion to be available?

Yes, we do.

**Comment by realvictorprm** [comment]
@S11001001 does [this](https://github.com/digital-asset/daml/pull/13460) cover the missing bit correctly?

**Comment by realvictorprm** [comment]
Oki, thanks folks! Will add it tomorrow and open another PR :)

Stephen Compall ***@***.***> schrieb am Di., 29. März 2022,
23:39:
